physics/checkpointer.hpp

```diff
@@ -0,0 +1,95 @@
+﻿
+#pragma once
+
+#include <functional>
+#include <tuple>
+
+#include "absl/synchronization/mutex.h"
+#include "base/not_null.hpp"
+#include "geometry/named_quantities.hpp"
+#include "quantities/quantities.hpp"
+
+namespace principia {
+namespace physics {
+namespace internal_checkpointer {
+
+using base::not_null;
+using geometry::Instant;
+using quantities::Time;
+
+// The checkpointer helps with compact serialization of timelines, i.e., classes
+// that associate some data with distinct instants.  The naïve implementation of
+// serialization for timelines would write all the pairs (time, data) but that
+// would potentially result in large saves that would be expensive to read and
+// write.
+// Instead, this class creates checkpoints that encapsulate all information
+// needed to reconstruct the timeline after a give point in time.  When
+// serializing a timeline, the pairs (time, data) are written up to the oldest
+// checkpoint, followed by the checkpoint itself.  When deserializing, the
+// the timeline may be reconstructed as needed based on the checkpoint.
+// Checkpoints must be created at regular intervals because they are dropped by
+// ForgetBefore: this ensures that there is always a sufficient old checkpoint
+// available the next time serialization is performed.
+// Logically checkpoints would serialize to/deserialize from a specific message,
+// but for historical reasons they just fill some fields of a message that may
+// contain other information.
+// This class is thread-safe.
+template<typename Message>
+class Checkpointer {
+ public:
+  // A function that reconstructs an object from a checkpoint.  It must return
```

Comment:
A function that reconstructs an object from a checkpoint **as a side effect**, I got confused looking for the object in the signature.

Alternatively, templatize on the object and pass it by non-const reference.

Reply:
I started with templatizing on the object, but it looked silly because it would take the pointer at construction and store it just to pass it back to the callbacks.  Incidentally, this schema would technically make it possible to checkpoint *several* objects together.

Beefed up the comment.